### ISSUE TITLE
Add implementation of RANDOM streaming percentile algorithm

### ISSUE THREAD

**Issue by psteitz** [Hipparchus-Math/hipparchus#8]
The [RANDOM](http://dimacs.rutgers.edu/~graham/pubs/papers/nquantiles.pdf) algorithm makes a nice complement to PSquarePercentile for streaming percentiles.  While it does not have a uniformly fixed bound on storage, storage is bounded for fixed quantile estimation error and grows very slowly with increases in precision.   It also allows any quantile to be estimated based on the data it stores, so e.g. getResult(quantile) or even getResult(quantile[]) methods are possible.  Finally, aggregation is straightforward.

**Comment by psteitz** [comment]
In commit d7686240fe7309bcf03f9a005fb40323731abd9a, I just added something to start with.  Comments / PRs / commits welcome.  Here are the things that I know of remaining to do:

1. Implement equals, hashcode, toString
2. Implement AggregatableStatistic.  The details will have to be worked out here; but the idea should be to do what increment does when there are no more buffers (find two among the lowest level and merge them) until the total number of buffers is less than the maximum.  Care will have to be taken when merging partially filled buffers (actually semantics of this has to be defined).
3. Clean up the constructor mess.  As usual, I have done a poor job adding lots of constructors.  Builder or other better ideas welcome. 
4. Add lots more tests - including Buffer and BufferMap
5. The test class is adapted from PSquarePercentile.  There is a test commented out both here and for PSquare for LogNormalDistribution that needs to be figured out.  Percentile itself returns infinite values for the test case, which makes no sense.
6. Add a description of the algorithm and how to use it to the User Guide.  
7. Might be good to add a static method that returns storage requirement for a given epsilon.

**Comment by psteitz** [comment]
Implemented AggregatableStatistic in ad93b8f, as well as a reduce method that just computes a percentile over a collection of aggregates (so does not actually merge all of the data).  Aggregation works as follows:

- If all of the aggregates are "half full" or less (meaning all have only level 0 data and are less than half capacity), the data is just combined.
- Otherwise, the combined set of buffers is reduced by repeatedly selecting two minimum level buffers and merging them, until the total number of buffers is within capacity.  This may require merging buffers of different levels.  To do that, the higher level buffer is retained and entries from the lower level buffer have lower probability of selection.
- Only RandomPercentile instances with the same buffer size can be aggregated.

In the commit above, I also added a RandomPercentile to StreamingStatistics.  This increases the memory requirement and will have some performance impact, so we may want to make this configurable.
